File: server/alertEngine.ts
/**
 * PRISM Early Warning Alert Engine
 *
 * Evaluates real longitudinal PAIMANA monthly observations and PRISM Risk assessments
 * to generate evidence-based early warnings.
 *
 * Core Principle:
 *   Alerts are NOT predictions ("This project will fail").
 *   They are evidence-based early warnings derived from observed data
 *   ("This project shows indicators requiring attention").
 *
 * Supported Alert Categories:
 *   1. Progress Stagnation: Physical progress changes very little across consecutive observations.
 *   2. Deteriorating Progress: Recent progress velocity is materially worse than earlier progress.
 *   3. Schedule Pressure: Target completion indicates increasing schedule pressure.
 *   4. Cost Escalation: Revised cost has increased relative to original cost.
 *   5. Physical-Financial Divergence: Expenditure progress and physical progress materially diverge.
 *   6. High/Critical Risk: Project crosses an important PRISM Risk Index threshold.
 */

import { Project, PaimanaObservation, PriorityTier } from '../src/types/index';

export type AlertType =
  | 'Progress Stagnation'
  | 'Deteriorating Progress'
  | 'Schedule Pressure'
  | 'Cost Escalation'
  | 'Physical-Financial Divergence'
  | 'High/Critical Risk';

export type AlertSeverity = 'CRITICAL' | 'HIGH' | 'MEDIUM';

export interface EarlyWarningAlert {
  id: string;
  projectId: string;
  projectName: string;
  projectCode: string;
  state: string;
  sector: string;
  implementingAgency?: string;
  alertType: AlertType;
  severity: AlertSeverity;
  evidence: string;
  detectedPeriod: string;
  recommendedAttention: string;
  riskScore: number | null;
  priorityScore: number | null;
  priorityTier: PriorityTier | null;
}

const MONTH_NAMES = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

function formatMonth(reportMonth?: string): string {
  if (!reportMonth) return 'Recent Observation';
  const parts = reportMonth.split('-');
  if (parts.length !== 2) return reportMonth;
  const m = parseInt(parts[1], 10);
  const y = parts[0];
  if (isNaN(m) || m < 1 || m > 12) return reportMonth;
  return `${MONTH_NAMES[m - 1]} ${y}`;
}

export class PRISMAlertEngine {
  /**
   * Evaluates a single project and its chronological monthly observations
   * to produce the single most impactful, non-repetitive early warning alert (if applicable).
   */
  public static evaluateProject(
    project: Project,
    observations?: PaimanaObservation[]
  ): EarlyWarningAlert | null {
    const trend = project.monthlyTrend || [];
    const len = trend.length;
    const obsList = observations || [];

    // 1. Progress Stagnation (Physical progress changed very little across consecutive observations)
    if (len >= 2 && project.physicalProgressPercent < 98) {
      let consecutiveStagnant = 0;
      for (let i = len - 1; i >= 1; i--) {
        const d = (trend[i].actualPercent ?? 0) - (trend[i - 1].actualPercent ?? 0);
        if (Math.abs(d) <= 0.2) {
          consecutiveStagnant++;
        } else {
          break;
        }
      }

      if (consecutiveStagnant >= 2) {
        const last = trend[len - 1];
        const firstStagnant = trend[len - 1 - consecutiveStagnant] || trend[len - 2];
        const totalDelta = (last.actualPercent ?? 0) - (firstStagnant.actualPercent ?? 0);
        const periodStr = `${formatMonth(firstStagnant.reportMonth || firstStagnant.month)} – ${formatMonth(last.reportMonth || last.month)}`;

        return {
          id: `alert-stag-${project.id}`,
          projectId: project.id,
          projectName: project.name,
          projectCode: project.code,
          state: project.state,
          sector: (project.derivedSector || project.sector) as string,
          implementingAgency: project.implementingAgency,
          alertType: 'Progress Stagnation',
          severity: consecutiveStagnant >= 3 ? 'CRITICAL' : 'HIGH',
          detectedPeriod: periodStr,
          evidence: `Physical progress changed by only ${totalDelta >= 0 ? '+' : ''}${totalDelta.toFixed(1)} percentage points across ${consecutiveStagnant} consecutive observation cycles (${firstStagnant.actualPercent}% to ${last.actualPercent}%).`,
          recommendedAttention: 'Conduct site execution review and verify contractor physical plant mobilization.',
          riskScore: project.riskScore,
          priorityScore: project.priorityScore ?? null,
          priorityTier: project.priorityTier ?? null
        };
      }
    }

    // 2. High/Critical Risk (Project crosses important PRISM Risk Index threshold)
    if (project.riskScore != null && project.riskScore >= 80) {
      return {
        id: `alert-risk-${project.id}`,
        projectId: project.id,
        projectName: project.name,
        projectCode: project.code,
        state: project.state,
        sector: (project.derivedSector || project.sector) as string,
        implementingAgency: project.implementingAgency,
        alertType: 'High/Critical Risk',
        severity: 'CRITICAL',
        detectedPeriod: formatMonth(project.lastUpdated),
        evidence: `Composite PRISM Risk Index reached ${project.riskScore}/100 (${project.riskTier}) confirmed by longitudinal monthly flash reports.`,
        recommendedAttention: 'Escalate to Ministry Project Monitoring Cell for monthly cabinet oversight review.',
        riskScore: project.riskScore,
        priorityScore: project.priorityScore ?? null,
        priorityTier: project.priorityTier ?? null
      };
    }

    // 3. Physical-Financial Divergence (Expenditure progress and physical progress materially diverge)
    const expPct = project.expenditurePctOfRevisedCost ?? (
      project.revisedCostCr > 0
        ? parseFloat(((project.cumulativeExpenditureCr / project.revisedCostCr) * 100).toFixed(1))
        : 0
    );
    const physPct = project.physicalProgressPercent ?? 0;
    const divergenceGap = parseFloat((expPct - physPct).toFixed(1));

    if (divergenceGap >= 20 && project.cumulativeExpenditureCr >= 10) {
      return {
        id: `alert-div-${project.id}`,
        projectId: project.id,
        projectName: project.name,
        projectCode: project.code,
        state: project.state,
        sector: (project.derivedSector || project.sector) as string,
        implementingAgency: project.implementingAgency,
        alertType: 'Physical-Financial Divergence',
        severity: divergenceGap >= 35 ? 'CRITICAL' : 'HIGH',
        detectedPeriod: formatMonth(project.lastUpdated),
        evidence: `Cumulative expenditure reached ${expPct.toFixed(1)}% of revised budget while physical progress stands at only ${physPct.toFixed(1)}% (${divergenceGap} pp divergence gap).`,
        recommendedAttention: 'Audit milestone-linked disbursement vouchers and contractor payment clearances.',
        riskScore: project.riskScore,
        priorityScore: project.priorityScore ?? null,
        priorityTier: project.priorityTier ?? null
      };
    }

    // 4. Cost Escalation (Revised cost has increased relative to original cost)
    if (project.costOverrunPercent >= 40 && project.originalCostCr >= 10) {
      return {
        id: `alert-cost-${project.id}`,
        projectId: project.id,
        projectName: project.name,
        projectCode: project.code,
        state: project.state,
        sector: (project.derivedSector || project.sector) as string,
        implementingAgency: project.implementingAgency,
        alertType: 'Cost Escalation',
        severity: project.costOverrunPercent >= 80 ? 'CRITICAL' : 'HIGH',
        detectedPeriod: formatMonth(project.lastUpdated),
        evidence: `Sanctioned cost increased by ${project.costOverrunPercent.toFixed(1)}% from ₹${project.originalCostCr.toLocaleString()} Cr original outlay to ₹${project.revisedCostCr.toLocaleString()} Cr revised outlay.`,
        recommendedAttention: 'Initiate Expenditure Finance Committee (EFC) revised cost sanction review.',
        riskScore: project.riskScore,
        priorityScore: project.priorityScore ?? null,
        priorityTier: project.priorityTier ?? null
      };
    }

    // 5. Schedule Pressure (Target completion indicates increasing schedule pressure)
    if ((project.urgency || 0) >= 85 && project.physicalProgressPercent < 90) {
      const remainingWork = (100 - project.physicalProgressPercent).toFixed(1);
      const targetDate = project.revisedCompletionDate || project.originalCompletionDate || 'Not specified';
      return {
        id: `alert-sched-${project.id}`,
        projectId: project.id,
        projectName: project.name,
        projectCode: project.code,
        state: project.state,
        sector: (project.derivedSector || project.sector) as string,
        implementingAgency: project.implementingAgency,
        alertType: 'Schedule Pressure',
        severity: 'HIGH',
        detectedPeriod: formatMonth(project.lastUpdated),
        evidence: `Target completion date (${targetDate}) indicates critical schedule pressure with ${remainingWork}% physical work remaining.`,
        recommendedAttention: 'Convene inter-agency taskforce to finalize revised critical path recovery plan.',
        riskScore: project.riskScore,
        priorityScore: project.priorityScore ?? null,
        priorityTier: project.priorityTier ?? null
      };
    }

    // 6. Deteriorating Progress (Recent progress velocity is materially worse than earlier progress)
    if (len >= 3) {
      const deltas: number[] = [];
      for (let i = 1; i < len; i++) {
        deltas.push((trend[i].actualPercent ?? 0) - (trend[i - 1].actualPercent ?? 0));
      }
      const recentDelta = deltas[deltas.length - 1];
      const earlyDeltas = deltas.slice(0, -1);
      const earlyAvg = earlyDeltas.reduce((a, b) => a + b, 0) / earlyDeltas.length;

      if (earlyAvg >= 1.0 && (earlyAvg - recentDelta) >= 1.5) {
        const last = trend[len - 1];
        return {
          id: `alert-decel-${project.id}`,
          projectId: project.id,
          projectName: project.name,
          projectCode: project.code,
          state: project.state,
          sector: (project.derivedSector || project.sector) as string,
          implementingAgency: project.implementingAgency,
          alertType: 'Deteriorating Progress',
          severity: recentDelta <= 0 ? 'HIGH' : 'MEDIUM',
          detectedPeriod: formatMonth(last.reportMonth || last.month),
          evidence: `Monthly progress pace slowed from an earlier average of ${earlyAvg.toFixed(1)} pp/mo to ${recentDelta.toFixed(1)} pp/mo in ${formatMonth(last.reportMonth || last.month)}.`,
          recommendedAttention: 'Investigate work front bottlenecks and verify field measurement certification.',
          riskScore: project.riskScore,
          priorityScore: project.priorityScore ?? null,
          priorityTier: project.priorityTier ?? null
        };
      }
    }

    return null;
  }

  /**
   * Generates a ranked, deduplicated list of meaningful alerts across the provided portfolio.
   */
  public static generateAlerts(
    projects: Project[],
    observationsMap?: Map<string, PaimanaObservation[]>
  ): EarlyWarningAlert[] {
    const alerts: EarlyWarningAlert[] = [];

    for (const project of projects) {
      const obsList = observationsMap?.get(project.id);
      const alert = PRISMAlertEngine.evaluateProject(project, obsList);
      if (alert) {
        alerts.push(alert);
      }
    }

    // Deterministic sort: CRITICAL first, then HIGH, then MEDIUM; within same severity, by priorityScore / riskScore
    const severityOrder: Record<AlertSeverity, number> = {
      CRITICAL: 0,
      HIGH: 1,
      MEDIUM: 2
    };

    return alerts.sort((a, b) => {
      const sevDiff = severityOrder[a.severity] - severityOrder[b.severity];
      if (sevDiff !== 0) return sevDiff;
      return (b.priorityScore ?? b.riskScore ?? 0) - (a.priorityScore ?? a.riskScore ?? 0);
    });
  }
}

File: backend/services/alert_engine.py
from typing import List, Optional, Dict
from backend.models.project import Project, PaimanaObservation, EarlyWarningAlert
from backend.models.common import AlertType, AlertSeverity

MONTH_NAMES = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']

def format_month(report_month: Optional[str]) -> str:
    if not report_month:
        return 'Recent Observation'
    parts = report_month.strip().split('-')
    if len(parts) != 2:
        return report_month
    try:
        y = parts[0]
        m = int(parts[1])
        if 1 <= m <= 12:
            return f"{MONTH_NAMES[m - 1]} {y}"
    except (ValueError, TypeError):
        pass
    return report_month

class PRISMAlertEngine:
    """
    Evidence-based Early Warning Alert Engine.
    Detects empirical operational indicators requiring attention across 6 alert categories.
    """
    @classmethod
    def evaluate_project(
        cls,
        project: Project,
        observations: Optional[List[PaimanaObservation]] = None
    ) -> Optional[EarlyWarningAlert]:
        trend = project.monthlyTrend or []
        length = len(trend)

        # 1. Progress Stagnation
        if length >= 2 and (project.physicalProgressPercent or 0.0) < 98.0:
            consecutive_stagnant = 0
            for i in range(length - 1, 0, -1):
                d = (trend[i].actualPercent or 0.0) - (trend[i - 1].actualPercent or 0.0)
                if abs(d) <= 0.2:
                    consecutive_stagnant += 1
                else:
                    break

            if consecutive_stagnant >= 2:
                last = trend[length - 1]
                first_stagnant = trend[length - 1 - consecutive_stagnant] if (length - 1 - consecutive_stagnant) >= 0 else trend[length - 2]
                total_delta = (last.actualPercent or 0.0) - (first_stagnant.actualPercent or 0.0)
                period_str = f"{format_month(first_stagnant.reportMonth or first_stagnant.month)} – {format_month(last.reportMonth or last.month)}"
                sign = '+' if total_delta >= 0 else ''

                return EarlyWarningAlert(
                    id=f"alert-stag-{project.id}",
                    projectId=project.id,
                    projectName=project.name,
                    projectCode=project.code,
                    state=project.state,
                    sector=project.derivedSector or project.sector,
                    implementingAgency=project.implementingAgency,
                    alertType='Progress Stagnation',
                    severity='CRITICAL' if consecutive_stagnant >= 3 else 'HIGH',
                    detectedPeriod=period_str,
                    evidence=f"Physical progress changed by only {sign}{total_delta:.1f} percentage points across {consecutive_stagnant} consecutive observation cycles ({first_stagnant.actualPercent}% to {last.actualPercent}%).",
                    recommendedAttention='Conduct site execution review and verify contractor physical plant mobilization.',
                    riskScore=project.riskScore,
                    priorityScore=project.priorityScore,
                    priorityTier=project.priorityTier
                )

        # 2. High/Critical Risk
        if project.riskScore is not None and project.riskScore >= 80:
            return EarlyWarningAlert(
                id=f"alert-risk-{project.id}",
                projectId=project.id,
                projectName=project.name,
                projectCode=project.code,
                state=project.state,
                sector=project.derivedSector or project.sector,
                implementingAgency=project.implementingAgency,
                alertType='High/Critical Risk',
                severity='CRITICAL',
                detectedPeriod=format_month(project.lastUpdated),
                evidence=f"Composite PRISM Risk Index reached {project.riskScore}/100 ({project.riskTier}) confirmed by longitudinal monthly flash reports.",
                recommendedAttention='Escalate to Ministry Project Monitoring Cell for monthly cabinet oversight review.',
                riskScore=project.riskScore,
                priorityScore=project.priorityScore,
                priorityTier=project.priorityTier
            )

        # 3. Physical-Financial Divergence
        rev_cost = project.revisedCostCr or 0.0
        exp_pct = project.expenditurePctOfRevisedCost
        if exp_pct is None:
            exp_pct = round(((project.cumulativeExpenditureCr or 0.0) / rev_cost) * 100.0, 1) if rev_cost > 0 else 0.0

        phys_pct = project.physicalProgressPercent or 0.0
        divergence_gap = round(exp_pct - phys_pct, 1)

        if divergence_gap >= 20.0 and (project.cumulativeExpenditureCr or 0.0) >= 10.0:
            return EarlyWarningAlert(
                id=f"alert-div-{project.id}",
                projectId=project.id,
                projectName=project.name,
                projectCode=project.code,
                state=project.state,
                sector=project.derivedSector or project.sector,
                implementingAgency=project.implementingAgency,
                alertType='Physical-Financial Divergence',
                severity='CRITICAL' if divergence_gap >= 35.0 else 'HIGH',
                detectedPeriod=format_month(project.lastUpdated),
                evidence=f"Cumulative expenditure reached {exp_pct:.1f}% of revised budget while physical progress stands at only {phys_pct:.1f}% ({divergence_gap:.1f} pp divergence gap).",
                recommendedAttention='Audit milestone-linked disbursement vouchers and contractor payment clearances.',
                riskScore=project.riskScore,
                priorityScore=project.priorityScore,
                priorityTier=project.priorityTier
            )

        # 4. Cost Escalation
        if (project.costOverrunPercent or 0.0) >= 40.0 and (project.originalCostCr or 0.0) >= 10.0:
            return EarlyWarningAlert(
                id=f"alert-cost-{project.id}",
                projectId=project.id,
                projectName=project.name,
                projectCode=project.code,
                state=project.state,
                sector=project.derivedSector or project.sector,
                implementingAgency=project.implementingAgency,
                alertType='Cost Escalation',
                severity='CRITICAL' if (project.costOverrunPercent or 0.0) >= 80.0 else 'HIGH',
                detectedPeriod=format_month(project.lastUpdated),
                evidence=f"Sanctioned cost increased by {project.costOverrunPercent:.1f}% from ₹{project.originalCostCr:,.0f} Cr original outlay to ₹{project.revisedCostCr:,.0f} Cr revised outlay.",
                recommendedAttention='Initiate Expenditure Finance Committee (EFC) revised cost sanction review.',
                riskScore=project.riskScore,
                priorityScore=project.priorityScore,
                priorityTier=project.priorityTier
            )

        # 5. Schedule Pressure
        if (project.urgency or 0) >= 85 and (project.physicalProgressPercent or 0.0) < 90.0:
            remaining_work = 100.0 - (project.physicalProgressPercent or 0.0)
            target_date = project.revisedCompletionDate or project.originalCompletionDate or 'Not specified'
            return EarlyWarningAlert(
                id=f"alert-sched-{project.id}",
                projectId=project.id,
                projectName=project.name,
                projectCode=project.code,
                state=project.state,
                sector=project.derivedSector or project.sector,
                implementingAgency=project.implementingAgency,
                alertType='Schedule Pressure',
                severity='HIGH',
                detectedPeriod=format_month(project.lastUpdated),
                evidence=f"Target completion date ({target_date}) indicates critical schedule pressure with {remaining_work:.1f}% physical work remaining.",
                recommendedAttention='Convene inter-agency taskforce to finalize revised critical path recovery plan.',
                riskScore=project.riskScore,
                priorityScore=project.priorityScore,
                priorityTier=project.priorityTier
            )

        # 6. Deteriorating Progress
        if length >= 3:
            deltas = [
                (trend[i].actualPercent or 0.0) - (trend[i - 1].actualPercent or 0.0)
                for i in range(1, length)
            ]
            recent_delta = deltas[-1]
            early_deltas = deltas[:-1]
            early_avg = sum(early_deltas) / (len(early_deltas) or 1)

            if early_avg >= 1.0 and (early_avg - recent_delta) >= 1.5:
                last = trend[length - 1]
                return EarlyWarningAlert(
                    id=f"alert-decel-{project.id}",
                    projectId=project.id,
                    projectName=project.name,
                    projectCode=project.code,
                    state=project.state,
                    sector=project.derivedSector or project.sector,
                    implementingAgency=project.implementingAgency,
                    alertType='Deteriorating Progress',
                    severity='HIGH' if recent_delta <= 0 else 'MEDIUM',
                    detectedPeriod=format_month(last.reportMonth or last.month),
                    evidence=f"Monthly progress pace slowed from an earlier average of {early_avg:.1f} pp/mo to {recent_delta:.1f} pp/mo in {format_month(last.reportMonth or last.month)}.",
                    recommendedAttention='Investigate work front bottlenecks and verify field measurement certification.',
                    riskScore=project.riskScore,
                    priorityScore=project.priorityScore,
                    priorityTier=project.priorityTier
                )

        return None

    @classmethod
    def generate_alerts(
        cls,
        projects: List[Project],
        observations_map: Optional[Dict[str, List[PaimanaObservation]]] = None
    ) -> List[EarlyWarningAlert]:
        alerts: List[EarlyWarningAlert] = []

        for project in projects:
            obs_list = observations_map.get(project.id) if observations_map else None
            alert = cls.evaluate_project(project, obs_list)
            if alert:
                alerts.append(alert)

        severity_order: Dict[str, int] = {
            'CRITICAL': 0,
            'HIGH': 1,
            'MEDIUM': 2
        }

        # Deterministic sort: CRITICAL first, then HIGH, then MEDIUM;
        # within same severity, by priorityScore / riskScore DESC
        def alert_sort_key(a: EarlyWarningAlert):
            score = a.priorityScore if a.priorityScore is not None else (a.riskScore if a.riskScore is not None else 0)
            return (severity_order.get(a.severity, 3), -score)

        alerts.sort(key=alert_sort_key)
        return alerts
